### ISSUE TITLE
Few random pieces of feedback

### ISSUE THREAD

**Issue by theScottyJam** [jsebrech/plainvanilla#5]
This website [was shared on reddit](https://www.reddit.com/r/SoftwareEngineering/comments/1flbjpz/plain_vanilla/), and I thought it was really cool, but there did seems to be a couple of places where it could be improved. I put together an entire response for reddit - then realized that it would probably make sense to share the same feedback here - so here ya go. Take it or leave it, but these are just my 2c (mostly copy-pasted).

---

* This website never dives very deep into a custom components lifecycle, and I wish it did, because it's more complicated than what you'd fine in traditional frameworks. In a traditional framework, you initialize a component, move it around as needed, then uninitialize it when you're done, then you never touch it again. Native DOM elements are a different story - You might detach an element from the DOM, but you might just turn around and re-attach it to the DOM somewhere else - never, at any point in time, do you send some kind of signal saying "I'm permanently done with you, go ahead and fully clean up", instead, it's up to each native element to unwire themselves whenever they're detached from the DOM (in preparation for potential garbage collection), then rewire themselves back up if/when they're re-attached to the DOM. Because custom HTML elements are modeled after native ones, they inherit the same, complicated lifecycle, which means component authors, in turn, inherit the same complicated lifecycle - something as simple as moving an element from one place to another will cause their "unwire everything" hook (`disconnectedCallback()`) to fire, immediately followed by their "wire everything up" hook (`connectedCallback()`). Because of this extra complexity, I know many users end up going "screw it, I'm just going to put my own lifecycle system on top of my components (e.g. by putting an unmount function on it)" - which is also a completely valid way to go. Unfortunately, none of this is discussed - what's more, multiple examples are actually buggy because they're not handling lifecycle properly - if you take one of the custom components from the examples, and try to move it in the DOM, you may end up with varying kind of buggy behavior - for example, the "advance component" example will cause empty image tags to be added to the custom element every time you move it, because it's treating `connectCallback()` like an on-mount that fires only once, when in reality it can fire many times. A more common bug is that everything just gets rerenders from scratch, which means input boxes will get cleared out, etc
* It kills me a little inside to see the author blindly interpolating variables into strings, that then are placed into `.innerHTML`. It would be nice to see some mention of how to avoid XSS vulnerabilities, and then properly build the examples to demonstrate how to escape HTML characters before interpolating. Even if the example isn't technically vulnerable to XSS, they're still somewhat buggy in that they're unable to render certain characters properly (such as `<`, `>`, and `&`).
* It has some odd advice on avoiding the shadow DOM. It mentions it impacts accessibility and SEO - something I've never heard of, and something that's not backed up in any way. It also mentions that there's an issue with content going unstyled for a moment while the stylesheet in the shadow DOM loads in - this used to be true, but isn't anymore if you use [constructable stylesheets](https://web.dev/articles/constructable-stylesheets).

**Comment by jsebrech** [comment]
Thanks for the feedback! I created a PR #6 to address your concerns. If that looks good to you I'll merge it.

Some thoughts on what I didn't address:

- On lifecycle: I think that a lot of tutorials get lost in the weeds when explaining the lifecycle, losing their audience, and that it sort of teaches itself as you start using custom elements. For that reason I tried to keep the lifecycle explanation to a minimum. I've tried to address the specific issues you mentioned in the PR. A lot of custom element implementations also spend time cleaning up when they don't need to. Browser event listeners don't prevent garbage collection, so mostly I have noticed that custom elements are garbage collected just fine if no cleanup is done.
- On shadow DOM: I think it's just not a good API to recommend people use by default. It adds complexity and introduces a number of issues (I have sourced those in the PR, but have personally ran into SEO and accessibility problems). Popular frameworks like React, Vue or Angular don't use shadow DOM, and I think for vanilla web dev similarly the advice should be that it is the exception rather than the rule. I do point out when shadow DOM is recommended in the tutorial, and in those circumstances I would use it myself.

**Comment by theScottyJam** [comment]
> React, Vue or Angular don't use shadow DOM

Angular and Vue actually both support the shadow-dom - it's just opt-in. In Angular, for example, when making your component, you just add `encapsulation: ViewEncapsulation.ShadowDom` and you get it. I can't help but wonder if they would have enabled the shadow DOM by default if they were built after the shadow DOM was standardized - I know many newer frameworks do.

But you're also right, it does add complexity. And I love the links you added about the shadow DOM - they were informative.

Overall, it looks good to me - thanks!

**Comment by jsebrech** [comment]
Thanks for taking a look at it!

If there's anything else you notice, please don't hesitate to report another issue.
